src/python/pants/backend/python/subsystems/python_native_code.py

```diff
@@ -129,3 +139,157 @@ def check_build_for_current_platform_only(self, targets):
+def select_cpp_toolchain_for_local_dist_compilation(python_native_code):
+  llvm_cpp_toolchain = yield Get(
+    LLVMCppToolchain, NativeToolchain, python_native_code.native_toolchain)
+  yield llvm_cpp_toolchain.cpp_toolchain
+
+
+class SetupPyNativeTools(datatype([
+    ('c_toolchain', CToolchain),
+    ('cpp_toolchain', CppToolchain),
+    ('platform', Platform),
+])):
+  """The native tools needed for a setup.py invocation.
+
+  This class exists because `SetupPyExecutionEnvironment` is created manually, one per target.
+  """
+
+
+# TODO: could this kind of @rule be automatically generated?
+@rule(SetupPyNativeTools, [Select(CToolchain), Select(CppToolchain), Select(Platform)])
+def get_setup_py_native_tools(c_toolchain, cpp_toolchain, platform):
+  yield SetupPyNativeTools(
+    c_toolchain=c_toolchain,
+    cpp_toolchain=cpp_toolchain,
+    platform=platform)
+
+
+class SetupRequiresSiteDir(datatype(['site_dir'])): pass
+
+
+# TODO: This could be formulated as an @rule if targets and `PythonInterpreter` are made available
+# to the v2 engine.
+def ensure_setup_requires_site_dir(reqs_to_resolve, interpreter, site_dir,
+                                   platforms=None):
+  if not reqs_to_resolve:
+    return None
+
+  setup_requires_dists = resolve_multi(interpreter, reqs_to_resolve, platforms, None)
+
+  # FIXME: there's no description of what this does or why it's necessary.
+  overrides = {
+    'purelib': site_dir,
+    'headers': os.path.join(site_dir, 'headers'),
+    'scripts': os.path.join(site_dir, 'bin'),
+    'platlib': site_dir,
+    'data': site_dir
+  }
+
+  # The `python_dist` target builds for the current platform only.
+  # FIXME: why does it build for the current platform only?
+  for obj in setup_requires_dists['current']:
+    wf = WheelFile(obj.location)
+    wf.install(overrides=overrides, force=True)
+
+  return SetupRequiresSiteDir(site_dir)
+
+
+# TODO: It might be pretty useful to have an Optional TypeConstraint.
+class SetupPyExecutionEnvironment(datatype([
+    # If None, don't set PYTHONPATH in the setup.py environment.
+    'setup_requires_site_dir',
+    # If None, don't execute in the toolchain environment.
+    'setup_py_native_tools',
+])):
+
+  _SHARED_CMDLINE_ARGS = {
+    'darwin': lambda: [
+      MIN_OSX_VERSION_ARG,
+      '-Wl,-dylib',
+      '-undefined',
+      'dynamic_lookup',
+    ],
+    'linux': lambda: ['-shared'],
+  }
+
+  def as_environment(self):
+    ret = {}
+
+    if self.setup_requires_site_dir:
+      ret['PYTHONPATH'] = self.setup_requires_site_dir.site_dir
+
+    # FIXME(#5951): the below is a lot of error-prone repeated logic -- we need a way to compose
+    # executables more hygienically. We should probably be composing each datatype's members, and
+    # only creating an environment at the very end.
+    native_tools = self.setup_py_native_tools
+    if native_tools:
+      # TODO: an as_tuple() method for datatypes could make this destructuring cleaner!
```

Comment:
I think this is actually a sign you should eliminate the middleman. See comment below RE SetupPyNativeTools rule.

Reply:
I have so far ended up keeping the `SetupPyExecutionEnvironment` for now, but removing all the rules associated with it. This is because the right environment is difficult to create simply by composing together other executable environments (for now) -- but I did expand the comment here and remove the `TODO` in 7ebfe51.